### PR TITLE
pytorch-bin: 1.10.0 -> 1.11.0

### PR DESCRIPTION
**Pull request by rehno-lindeque**
###### Description of changes

This is still a work in progress (needs testing, etc)

1.11 Release of PyTorch:

* [Announcement](https://pytorch.org/blog/pytorch-1.11-released/)
* [Pytorch 1.11.0 release notes](https://github.com/pytorch/pytorch/releases/tag/v1.11.0)

Nix changes:

- [X] Bumped pytorch as well as pytorch-bin to 1.11
- [X] Added darwin dowloads to the prefetch script
- [X] Added cp310 to the prefetch script for linux
- [x] Enable pytorch-bin for (linux, python 3.10)

TODO

- [ ] Check: are torchdata, functorch automatically bundled into the release?
- [ ] Tensorboard dependency doesn't build on python 3.10 due to https://github.com/NixOS/nixpkgs/commit/a43876176fe7df59c793db8b367df26bec8b7084
  * See https://github.com/NixOS/nixpkgs/pull/164730
- [ ] ~CUDA appears to need updating~
* See https://github.com/NixOS/nixpkgs/pull/164338/files/2000db2e332b552c3a7237b8fafb60be2c6687d2#r830322195
```
-- Found CUDA: /nix/store/rclmj9izhxwhgwrgjk97mka3ndm4bzyn-cudatoolkit-10.1.243 (found version "10.1") 
-- Caffe2: CUDA detected: 10.1
-- Caffe2: CUDA nvcc is: /nix/store/rclmj9izhxwhgwrgjk97mka3ndm4bzyn-cudatoolkit-10.1.243/bin/nvcc
-- Caffe2: CUDA toolkit directory: /nix/store/rclmj9izhxwhgwrgjk97mka3ndm4bzyn-cudatoolkit-10.1.243
CMake Error at cmake/public/cuda.cmake:42 (message):
  PyTorch requires CUDA 10.2 or above.
Call Stack (most recent call first):
  cmake/Dependencies.cmake:1191 (include)
  CMakeLists.txt:653 (include)
```

- [ ] Needs building, testing, etc (in progress)

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/pytorch/binary-hashes.nix
Comment (rehno-lindeque) [inline pkgs/development/python-modules/pytorch/binary-hashes.nix:50]: I kept the old pytorch 1.10 binary hashes around, but I'm happy to delete them if that is the correct workflow.

Comment (junjihashimoto) [reply pkgs/development/python-modules/pytorch/binary-hashes.nix:50]: I also think you can delete previous pytorch 1.10.0 binary hashes.

File: pkgs/development/python-modules/pytorch/default.nix
Comment (junjihashimoto) [inline pkgs/development/python-modules/pytorch/default.nix:33]: The binary uses cuda11.3.

**Review by junjihashimoto — CHANGES_REQUESTED** [review]
Thank you for creating a PR!
torchvision-bin and torchaudio-bin also depend on this.
We should include them in another commit or require another PR.

**Comment by junjihashimoto** [comment]
Did prefetch.sh work fine?
When I tried it with libtorch, it stopped for some reason. 
The problem is not well understood, but it is reproducible.

**Comment by junjihashimoto** [comment]
Technically, the binary does not need nvcc with cudatoolkit.

**Comment by rehno-lindeque** [comment]
> Did prefetch.sh work fine? When I tried it with libtorch, it stopped for some reason. The problem is not well understood, but it is reproducible.

I'll double check it (I ran it in two steps before, so possible I made a mistake somewhere in updating the script)

File: pkgs/development/python-modules/pytorch/default.nix
Comment (rehno-lindeque) [reply pkgs/development/python-modules/pytorch/default.nix:33]: Thanks cuda that is passed in should probably be 11.3 by default then. 

My take that 10.2.0 is the minimum was based on a CMake error I got, but I haven't properly investigated if that's truly the lower bound.
```
-- Found CUDA: /nix/store/rclmj9izhxwhgwrgjk97mka3ndm4bzyn-cudatoolkit-10.1.243 (found version "10.1") 
-- Caffe2: CUDA detected: 10.1
-- Caffe2: CUDA nvcc is: /nix/store/rclmj9izhxwhgwrgjk97mka3ndm4bzyn-cudatoolkit-10.1.243/bin/nvcc
-- Caffe2: CUDA toolkit directory: /nix/store/rclmj9izhxwhgwrgjk97mka3ndm4bzyn-cudatoolkit-10.1.243
CMake Error at cmake/public/cuda.cmake:42 (message):
  PyTorch requires CUDA 10.2 or above.
Call Stack (most recent call first):
  cmake/Dependencies.cmake:1191 (include)
  CMakeLists.txt:653 (include)
```


(Btw this check doesn't seem to trigger for cudnn_cudatoolkit which is currently against 10.1.0 - I'm looking at fixing via https://github.com/NixOS/nixpkgs/pull/164820)

**Comment by junjihashimoto** [comment]
I think the title and commit should be 'python3Packages.pytorch-bin: 1.10.2 -> 1.11.0'.
pytorch is another derivation.

**Comment by junjihashimoto** [comment]
I was a little misunderstood. This includes both full build and binary derivations.

File: pkgs/development/python-modules/pytorch/default.nix
Comment (junjihashimoto) [reply pkgs/development/python-modules/pytorch/default.nix:33]: The problem was that I was a little misunderstood. I think this is okay.

File: pkgs/development/python-modules/pytorch/prefetch.sh
Comment (junjihashimoto) [inline pkgs/development/python-modules/pytorch/prefetch.sh:20]: https://download.pytorch.org/whl/cpu/torch-1.11.0-cp310-none-macosx_11_0_arm64.whl
cp310 of macosx exists.

File: pkgs/development/python-modules/pytorch/default.nix
Comment (samuela) [reply pkgs/development/python-modules/pytorch/default.nix:33]: To unblock you on the CUDA/cuDNN issue as quickly as possible, I would suggest just overriding with the specific versions you need here

https://github.com/NixOS/nixpkgs/blob/83e8ce412b4c7561b28acc924bd46a4bd08c2351/pkgs/top-level/python-packages.nix#L8312-L8314

a la

https://github.com/NixOS/nixpkgs/blob/83e8ce412b4c7561b28acc924bd46a4bd08c2351/pkgs/top-level/python-packages.nix#L4203-L4215

There's no need to rely too heavily on the `cudatoolkit` and `cudnn` aliases if they aren't currently pointing to the versions that you need.

File: pkgs/development/python-modules/pytorch/bin.nix
Comment (samuela) [inline pkgs/development/python-modules/pytorch/bin.nix:32]: Once/before this PR is ready please add a comment accompanying this change explaining why this combo is disabled with a link to the relevant PR/issue.

File: pkgs/development/python-modules/pytorch/prefetch.sh
Comment (rehno-lindeque) [reply pkgs/development/python-modules/pytorch/prefetch.sh:20]: Sorry about that, I guess I missed it

File: pkgs/development/python-modules/pytorch/bin.nix
Comment (rehno-lindeque) [reply pkgs/development/python-modules/pytorch/bin.nix:32]: I guess this is not correct since a macosx binary does exist apparently https://github.com/NixOS/nixpkgs/pull/164712/files/2c200e7142f4644bca9c29bc49525cbd53c28450#r830446187

Not sure how I missed it

**Comment by rehno-lindeque** [comment]
> I think the title and commit should be 'python3Packages.pytorch-bin: 1.10.2 -> 1.11.0'. pytorch is another derivation.

Happy to squash and rename to `python3Packages.pytorch: 1.10.2 -> 1.11.0' when this is ready (if that format looks correct)

**Comment by samuela** [comment]
>  CUDA appears to need updating

cc @NixOS/cuda-maintainers 

This failure is due to the CUDA version being picked up from the `cudnn` package instead of `cudatoolkit`. It's a bit messy but the latest cuDNN version supported by pytorch v1.10.2 was cudnn v7.6.5 AFAICT. And cudnn 7.6.5 supported at most CUDA 10.1 officially (it seems to unofficially work with CUDA 10.2 as well). In nix we try to enforce the official version constraints as much as possible.

Long story short, try the following in python-packages.nix:
```nix
  pytorch = callPackage ../development/python-modules/pytorch {
    cudaSupport = pkgs.config.cudaSupport or false;
    cudatoolkit = pkgs.cudatoolkit_11;
    cudnn = pkgs.cudnn_8_3_cudatoolkit_11;
  };
```
if that doesn't work, we'll need to figure out the latest version of cuDNN that pytorch supports compiling against. cuDNN v8.3.5 is currently the latest upstream and the latest packaged in nixpkgs. But there are [other options as well](https://cs.github.com/NixOS/nixpkgs/blob/0c3a726c4786f00b440de8158a2b9bfda79e3e3d/pkgs/development/libraries/science/math/cudnn/default.nix?q=cudnn). Don't hesitate to reach out with any questions! I guarantee that the CUDA/cuDNN versions you need are packaged... just a matter of figuring out what works.

**Comment by mweinelt** [comment]
Already on staging.

**Comment by rehno-lindeque** [comment]
Ah I missed comments here - glad to see others are beating me to it

**Comment by junjihashimoto** [comment]
@mweinelt 
Awesome!
Could you tell me where is the job of pytorch-1.11?
It looks like it's not in hydra's job.
https://hydra.nixos.org/search?query=pytorch

**Comment by samuela** [comment]
@junjihashimoto nixpkgs-unstable has been failing for the last 7 days, so we likely won't be able to see any builds until that is cleared up. Check out https://status.nixos.org for more info.

File: pkgs/development/python-modules/pytorch/default.nix
Comment (rehno-lindeque) [reply pkgs/development/python-modules/pytorch/default.nix:33]: I've removed pytorch source build from this PR now since it's already merged elsewhere :+1:

Comment (junjihashimoto) [inline pkgs/development/python-modules/pytorch/default.nix:33]: It is an assertion of pytorch derivation, not pytorch-bin. I don't think it's necessary to put it back.

**Review by junjihashimoto — APPROVED** [review]
It doesn't include torchvision-bin and torchaudio-bin, but pytorch-bin itself is fine.

File: pkgs/development/python-modules/pytorch/default.nix
Comment (rehno-lindeque) [reply pkgs/development/python-modules/pytorch/default.nix]: I meant to drop this, thanks

**Review by junjihashimoto — APPROVED** [review]
LGTM!  
(I prefer to commit cleanup because these depend on pytorch-bin.)

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 164712` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>tts</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>python310Packages.pytorch-bin</li>
    <li>python310Packages.torchaudio-bin</li>
    <li>python310Packages.torchvision-bin</li>
    <li>python39Packages.pytorch-bin</li>
    <li>python39Packages.torchaudio-bin</li>
    <li>python39Packages.torchvision-bin</li>
  </ul>
</details>

**Comment by samuela** [comment]
IIUC this PR bumps pytorch-bin, torchaudio-bin, and torchvision-bin. Do the torchaudio and torchvision source builds need to be updated as well?

**Comment by junjihashimoto** [comment]
The torchaudio and torchvision source builds generate different derivations than *-bin.
When we update pytorch source build not binary, the torchaudio and torchvision source builds need to be updated.

**Comment by junjihashimoto** [comment]
coqui-trainer depends on pytorch not pytorch-bin.
tts depends on pytorch-bin not pytorch.

**Comment by junjihashimoto** [comment]
There is no source build for torchaudio.

**Comment by junjihashimoto** [comment]
tts depends on coqui-trainer.
coqui-trainer depends on tensorboardx.
tensorboardx throws error on pytorch-1.11.
coqui-trainer should use not tensorboardx but torch.utils.tensorboard.

**Comment by junjihashimoto** [comment]
https://github.com/rehno-lindeque/nixpkgs/pull/1/files

File: pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch
Comment (mweinelt) [inline pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch:24]: Can we please contact the upstream about such a change?

File: pkgs/development/python-modules/coqui-trainer/default.nix
Comment (mweinelt) [inline pkgs/development/python-modules/coqui-trainer/default.nix:34]: I did this dance twice now, it is annyoing, because whenever a dependency switches between bin and src I have to follow. :facepalm:

File: pkgs/tools/audio/tts/tts.patch
Comment (mweinelt) [inline pkgs/tools/audio/tts/tts.patch:9]: We don't carry patches for minimal changes like this. Use `sed` in `postPatch` if you must.

**Comment by rehno-lindeque** [comment]
Is `pytorch-bin` important enough to have a `pytorch-bin_1_10`? I'm happy to keep at it, just curious if others would prefer more granular PRs for rolling pytorch ecosystem forward.

File: pkgs/tools/audio/tts/tts.patch
Comment (junjihashimoto) [reply pkgs/tools/audio/tts/tts.patch:9]: I've checked tts with git grep but it does not use tensorboardX.

File: pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch
Comment (junjihashimoto) [reply pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch:24]: https://github.com/coqui-ai/Trainer/pull/21

File: pkgs/tools/audio/tts/tts.patch
Comment (mweinelt) [reply pkgs/tools/audio/tts/tts.patch:9]: https://github.com/coqui-ai/TTS/issues/1433

File: pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch
Comment (mweinelt) [reply pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch:24]: Thank you!

File: pkgs/tools/audio/tts/tts.patch
Comment (mweinelt) [reply pkgs/tools/audio/tts/tts.patch:9]:
```patch
diff --git a/pkgs/tools/audio/tts/default.nix b/pkgs/tools/audio/tts/default.nix
index a7b0212a7b0..a15a29f9d57 100644
--- a/pkgs/tools/audio/tts/default.nix
+++ b/pkgs/tools/audio/tts/default.nix
@@ -57,6 +57,7 @@ python.pkgs.buildPythonApplication rec {
         ''-e 's/${package}.*[<>=]+.*/${package}/g' \''
       ) relaxedConstraints)}
     requirements.txt
+    sed -i '/tensorboardX/d' requirements.txt
   '';
 
   nativeBuildInputs = with python.pkgs; [
@@ -84,7 +85,6 @@ python.pkgs.buildPythonApplication rec {
     pyworld
     scipy
     soundfile
-    tensorboardx
     tensorflow
     torchaudio-bin
     tqdm

```

**Comment by samuela** [comment]
> Is `pytorch` and/or `pytorch-bin` important enough to have a `pytorch_1_10` and/or `pytorch-bin_1_10`? I'm happy to keep at it, just curious if others would prefer more granular PRs for rolling pytorch ecosystem forward.

My 2c is that I think it's best to avoid the complexity associated with multiple versions if possible, but I'm not a pytorch maintainer so I'll leave that to @junjihashimoto @teh @thoughtpolice and @tscholak.

**Review by junjihashimoto — APPROVED** [review]
Keeping multiple versions may just put off the problem.
We may need more than one version for another reason.
I'm not sure if we should manage the set of packages or follow conda or anything else.

**Comment by rehno-lindeque** [comment]
> There is no source build for torchaudio.

I had a draft PR for torchaudio source that I never quite got to finishing up properly. https://github.com/NixOS/nixpkgs/pull/160210 with some test dependencies https://github.com/NixOS/nixpkgs/pull/160206/files and https://github.com/NixOS/nixpkgs/pull/160197

It had a lot of failing tests that took crazy times to run, so it was a bit difficult to verify. (I don't use torchaudio myself either, and really don't want to be a maintainer).

Anyway, if anyone is so inclined, feel free to reappropriate it

**Comment by junjihashimoto** [comment]
I also tried it, but I couldn't complete many dependencies other than python.

File: pkgs/tools/audio/tts/tts.patch
Comment (mweinelt) [reply pkgs/tools/audio/tts/tts.patch]: Can you please fixup that commit and squash it into the previous one?

**Comment by rehno-lindeque** [comment]
Squashed superfluous commits & rebased

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 164712` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python310Packages.coqui-trainer</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>python310Packages.pytorch-bin</li>
    <li>python310Packages.torchaudio-bin</li>
    <li>python310Packages.torchvision-bin</li>
    <li>python39Packages.coqui-trainer</li>
    <li>python39Packages.pytorch-bin</li>
    <li>python39Packages.torchaudio-bin</li>
    <li>python39Packages.torchvision-bin</li>
    <li>tts</li>
  </ul>
</details>

**Comment by samuela** [comment]
coqui-trainer isn't building on python310 but I don't view that as a blocker for merging this. Lots of packages still failing on python310.
```
error: builder for '/nix/store/1x2jysbhgkd22dpaaa5rwq0z7zjwpfxr-python3.10-coqui-trainer-0.0.5.drv' failed with exit code 1;
       last 10 log lines:
       >     File "/nix/store/5gdg9a31yh18gsky303g6id7w2da9pvk-python3.10-setuptools-57.2.0/lib/python3.10/site-packages/setuptools/build_meta.py", line 258, in run_setup
       >       super(_BuildMetaLegacyBackend,
       >     File "/nix/store/5gdg9a31yh18gsky303g6id7w2da9pvk-python3.10-setuptools-57.2.0/lib/python3.10/site-packages/setuptools/build_meta.py", line 150, in run_setup
       >       exec(compile(code, __file__, 'exec'), locals())
       >     File "setup.py", line 36, in <module>
       >       raise RuntimeError(
       >   RuntimeError: Coqui-Trainer requires python >= 3.6 and <=3.10 but your Python version is 3.10.2 (main, Jan 13 2022, 19:06:22) [GCC 10.3.0]
       >   Preparing metadata (pyproject.toml) ... error
       > WARNING: Discarding file:///build/source. Command errored out with exit status 1: /nix/store/7mv9crg6y9bxgn39ynhkkwi3lhhsqhaj-python3-3.10.2/bin/python3.10 /nix/store/gximmp563q6s55srwds37hvx5mhacsxy-python3.10-pip-21.3.1/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /build/tmp3s5v5if6 Check the logs for full command output.
       > ERROR: Command errored out with exit status 1: /nix/store/7mv9crg6y9bxgn39ynhkkwi3lhhsqhaj-python3-3.10.2/bin/python3.10 /nix/store/gximmp563q6s55srwds37hvx5mhacsxy-python3.10-pip-21.3.1/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /build/tmp3s5v5if6 Check the logs for full command output.
```

File: pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch
Comment (samuela) [reply pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch:24]: Now that we have that PR open let's use `fetchpatch` here.

File: pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch
Comment (samuela) [inline pkgs/development/python-modules/coqui-trainer/conqui-trainer.patch:1]: Like @mweinelt, I generally prefer sed/substituteInPlace/etc instead of checked-in patch files. Patch files are much harder to update once a new version comes around. Using sed/substituteInPlace usually has a longer shelf-life.

But more importantly this patch file is no longer necessary since we can `fetchpatch` the upstream PR https://github.com/coqui-ai/Trainer/pull/21.

**Review by samuela — CHANGES_REQUESTED** [review]
let's use fetchpatch and then I think this PR is good to go!

**Comment by mweinelt** [comment]
Let's add a disable for python3.10 on coqui-trainer with a reference to https://github.com/coqui-ai/Trainer/issues/22.

```
  disabled = pythonAtLeast "3.10"; # https://github.com/coqui-ai/Trainer/issues/22
```

I agree with the remarks made by @samuela, patches need to go, fetchpatch is the way.

**Review by samuela — APPROVED** [review]
Ok LGTM! I'll go ahead and merge tomorrow unless anyone objects

**Comment by rehno-lindeque** [comment]
Thanks for all the help!

**Comment by samuela** [comment]
Great work, and thank you for your persistence with this @rehno-lindeque! It's not easy getting PRs through for some of these larger packages but it's absolutely crucial work!

**Pull request merged**
